mbuild/lattice.py

```diff
@@ -0,0 +1,401 @@
+                                     'equal to the dimension specified: {}'
+                                     .format(lst[1], dimension))
+        else:
+            raise TypeError('Incorrect Type: "basis_vectors" is of type {}, '
+                            'not type <list> or <tuple>. Please review '
+                            'the documentation for proper format.'
+                            .format(type(basis_vectors)))
+
+        if not isinstance(basis_vectors, defaultdict):
+            overlap_dict = defaultdict(list)
+            for the_id, vector in basis_vectors:
+                num_iterations = 3
+                if dimension == 3:
+                    for x in range(num_iterations):
+                        for y in range(num_iterations):
+                            for z in range(num_iterations):
+                                tmpx = vector[0] + x
+                                tmpy = vector[1] + y
+                                tmpz = vector[2] + z
+                                tmp_tuple = tuple((tmpx, tmpy, tmpz))
+                                value = tuple((the_id, vector))
+                                overlap_dict[tmp_tuple].append(value)
+
+                elif dimension == 2:
+                    for x in range(num_iterations):
+                        for y in range(num_iterations):
+                            tmpx = vector[0] + x
+                            tmpy = vector[1] + y
+                            tmp_tuple = tuple((tmpx, tmpy))
+                            value = tuple((the_id, vector))
+                            overlap_dict[tmp_tuple].append(value)
+
+                else:
+                    for x in range(num_iterations):
+                        tmpx = vector[0] + x
+                        tmp_tuple = tuple((tmpx))
+                        value = tuple((the_id, vector))
+                        overlap_dict[tmp_tuple].append(value)
+
+            # now review for overlap
+            for key, val in overlap_dict.items():
+                if len(val) > 1:
+                    raise ValueError('Overlapping Basis Vectors: Basis '
+                                     'vectors overlap when the unit cell is '
+                                     'expanded to {}. This is an incorrect '
+                                     'perfect lattice. The offending '
+                                     'vectors are: {}'
+                                     .format(key, val))
+        basis_dict = defaultdict(list)
+        if not isinstance(basis_vectors, defaultdict):
+            for key, val in basis_vectors:
+                basis_dict[key].append(val)
+        else:
+            basis_dict = deepcopy(basis_vectors)
+
+        self.dimension = dimension
+        self.lattice_vectors = lattice_vectors
+        self.lattice_spacings = lattice_spacings
+        self.basis_vectors = basis_dict
+
+    def populate(self, x=None, y=None, z=None, compound_dict=None):
```

Comment:
I would put compounddict first since I would consider that a more common use case.